File: 프로그래머스/lv2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    
    for i in range(len(dungeons), 0, -1):
        case = permutations(sorted(dungeons, reverse=True), i)
        for j in case:
            tired = k
            count = 0
            for dungeon in j:
                if dungeon[0] > tired:
                    break
                else:
                    tired -= dungeon[1]
                    count+=1
            if count == i:
                return i

File: 프로그래머스/2/87946. 피로도/피로도.js
const getPermutation = (arr, selectNumber) => {
  const result = [];
  if (selectNumber === 1) return arr.map((value) => [value]);

  arr.forEach((fixed, index, origin) => {
    const rest = [...origin.slice(0, index), ...origin.slice(index + 1)];
    const permutations = getPermutation(rest, selectNumber - 1);
    const attached = permutations.map((permutation) => [fixed, ...permutation]);
    result.push(...attached);
  });

  return result;
};

const clearDungeonCount = (myFatigue, permutations) => {
    const counts = []
    permutations.forEach((permutation) => {
        let fatigue = myFatigue
        let count = 0
        for([minFatigue, exhaustFatigue] of permutation) {
            if(minFatigue <= fatigue) {
                fatigue -= exhaustFatigue
                count+=1
            }
        }
        counts.push(count)
    })
    
    return Math.max(...counts)
}

function solution(k, dungeons) {
    const permutations = getPermutation(dungeons, dungeons.length);
    
    return clearDungeonCount(k, permutations);
}